### PR TITLE
Allow patrol areas to be single tiles

### LINKED ISSUE #681: Allow More Specific Patrol Area
I don't know what people's opinions on this would be, but I've been playing RCT1 lately and it really bugs me that I can't adjust the patrol area one square at a time.  I want to be able to restrict each handyman to a very specific area, but it's hard because you can only set patrol area in 4x4 squares at a time.

### PR DESCRIPTION
**Pull request by IntelOrca**
Change the patrol areas to be stored as a sorted list of points grouped in fixed cells for each staff entity. The complexity will be a mixed O(1) and O(log n) search.

Closes #681

image

### CODE REVIEW

File: src/openrct2/rct1/S4Importer.cpp [hunk 1]
```diff
@@ -1360,7 +1360,13 @@ namespace RCT1
...
                     int32_t y = val & 0x3E0;
                     y <<= 2;
-                    staffmember->SetPatrolArea({ x, y }, true);
+                    for (int32_t offsetY = 0; offsetY < 4 * COORDS_XY_STEP; offsetY += COORDS_XY_STEP)
+                    {
+                        for (int32_t offsetX = 0; offsetX < 4 * COORDS_XY_STEP; offsetX += COORDS_XY_STEP)
+                        {
+                            staffmember->SetPatrolArea({ x + offsetX, y + offsetY }, true);
+                        }
+                    }
```
Comment (pizza2004) [inline src/openrct2/rct1/S4Importer.cpp:1369]: This could be turned into a function to be used for both S4 and S6, right?

File: src/openrct2/entity/Staff.h [hunk 2]
```diff
@@ -22,7 +22,36 @@ constexpr size_t STAFF_PATROL_AREA_SIZE = (STAFF_PATROL_AREA_BLOCKS_PER_LINE * S
...
+    std::array<Cell, NumCells> Areas;
+    size_t TileCount{};
+
+    const Cell* GetCell(TileCoordsXY pos) const;
+    Cell* GetCell(TileCoordsXY pos);
+
+public:
+    bool IsEmpty() const;
+    void Clear();
+    bool Get(TileCoordsXY pos) const;
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Staff.h:48]: all TileCoordsXY and CoordsXY should be passed by const reference.

File: src/openrct2/entity/Staff.cpp [hunk 3]
```diff
@@ -68,6 +68,116 @@ colour_t gStaffSecurityColour;
...
+static bool CompareTileCoordsXY(const TileCoordsXY& lhs, const TileCoordsXY& rhs)
+{
+    if (lhs.y == rhs.y)
+        return lhs.x < rhs.x;
+    return lhs.y < rhs.y;
+}
+
+const PatrolArea::Cell* PatrolArea::GetCell(TileCoordsXY pos) const
+{
+    return const_cast<PatrolArea*>(this)->GetCell(pos);
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Staff.cpp:80]: why is this required? Just provide a const GetCell

File: src/openrct2/entity/Staff.h [hunk 4]
```diff
@@ -22,7 +22,36 @@ constexpr size_t STAFF_PATROL_AREA_SIZE = (STAFF_PATROL_AREA_BLOCKS_PER_LINE * S
 
 struct PatrolArea
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Staff.h:23]: move this to a new file staff.h and staff.cpp is already too big.

File: src/openrct2/park/ParkFile.cpp [hunk 5]
```diff
@@ -2015,64 +2015,23 @@ namespace OpenRCT2
...
-        if (cs.GetMode() == OrcaStream::Mode::WRITING)
+        if (cs.GetMode() == OrcaStream::Mode::WRITING && entity.PatrolInfo != nullptr)
         {
-            patrolArea = GetPatrolArea(entity);
+            patrolArea = entity.PatrolInfo->ToVector();
         }
         cs.ReadWriteVector(patrolArea, [&cs](TileCoordsXY& value) { cs.ReadWrite(value); });
         if (cs.GetMode() == OrcaStream::Mode::READING)
         {
-            SetPatrolArea(entity, patrolArea);
```
Comment (duncanspumpkin) [inline src/openrct2/park/ParkFile.cpp:2075]: This used to only create a PatrolArea if there was a patrolArea it is now always creating one even when it isn't required (zero sized vector)

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/entity/PatrolArea.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+bool PatrolArea::Get(const TileCoordsXY& pos) const
+{
+    auto* area = GetCell(pos);
+    if (area == nullptr)
+        return false;
+
+    auto it = std::lower_bound(area->SortedTiles.begin(), area->SortedTiles.end(), pos, CompareTileCoordsXY);
+    auto found = it != area->SortedTiles.end() && *it == pos;
```
Comment (CookiePLMonster) [inline src/openrct2/entity/PatrolArea.cpp:60]: IMO this and the use above should be a separate function, much like this now-removed `binary_find` example from cppreference:
https://sites.radford.edu/~acm/midatl/docs/cppreference/w/cpp/algorithm/lower_bound.html

Comment (IntelOrca) [reply src/openrct2/entity/PatrolArea.cpp:60]: Only for `get`, since `set` still needs to know where to insert the item. Maybe when there is a second use, we could make the trouble of a shared helper for it.

**Comment by oli414** [comment]
This is great! I do think that the introduction for this calls for some additional control:

image

Covering a section of a park without the ability to change the "brush size" would be incredibly tedious, and especially easy to miss a tile which would cause major issues.

Proposals:
A) Clicking the patrol area button makes the "brush size" window appear, clicking the patrol area button again, or closing the brush size window cancels setting the patrol area

B) Have a default brush size of 4x4, but have a new dropdown option for the patrol area button for "Precise Control" that has the 1x1 brush.

**Comment by Legotown55** [comment]
When will this be avaoble

**Comment by Legotown55** [comment]
When will 16740 updated be avalble

**Comment by karst** [comment]
> When will 16740 updated be avalble
Please stop spamming. It will be available once this is ready. It's already a PR, so the devs will playtest it before it gets added. This means it could take a little, or might be added really soon. Either way: Please stop spamming.

**Comment by Legotown55** [comment]
Sorry

**Comment by IntelOrca** [comment]
@oli414 Agreed, although the scope of this is growing too much. It would be better to leave this for a follow up PR, and it would be nice for someone else to do so I can focus on the plugin PRs.

**Comment by oli414** [comment]
@IntelOrca Totally agree, I'm just a bit concerned that users will have trouble with the single tile patrol area without additional tools. Perhaps its possible to make this PR about making it technically possible without changing the way users interact with it for now?

**Comment by duncanspumpkin** [comment]
> @IntelOrca Totally agree, I'm just a bit concerned that users will have trouble with the single tile patrol area without additional tools. Perhaps its possible to make this PR about making it technically possible without changing the way users interact with it for now?

Nah just leave it for a future pr. It's not going to take long to do we are talking perhaps a few days where the behaviour will be a bit harder.

**Comment by IntelOrca** [comment]
I have added a tool window now so you can change the size of the patrol area tool.

**Comment by karst** [comment]
> Your pacch for openrct2 has failed

Dev's already get notified about this ;) No need to point it out. I think IO might already be working on fixing it. If you wish to devtalk, please join the discord instead

File: src/openrct2-ui/windows/PatrolArea.cpp [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            gMapSelectPositionB = posB;
+            map_invalidate_selection_rect();
+        }
+    }
+
+    void OnToolAbort(rct_widgetindex widgetIndex) override
+    {
+        hide_gridlines();
+        ClearPatrolAreaToRender();
+        gfx_invalidate_screen();
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/PatrolArea.cpp:193]: Should this window auto close on tool abort?

Comment (IntelOrca) [reply src/openrct2-ui/windows/PatrolArea.cpp:193]: It does

File: src/openrct2-ui/windows/PatrolArea.cpp [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+    std::optional<CoordsXY> GetBestCoordsFromPos(const ScreenCoordsXY& pos)
+    {
+        auto coords = footpath_get_coordinates_from_pos(pos, nullptr, nullptr);
+        return coords.IsNull() ? std::nullopt : std::make_optional(coords);
+    }
+};
+
+rct_window* WindowPatrolAreaOpen(EntityId staffId)
+{
+    auto current = reinterpret_cast<PatrolAreaWindow*>(window_find_by_class(WC_PATROL_AREA));
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/PatrolArea.cpp:282]: i don't quite see the reason for this code. It basically does focus or create then does focus or create again. Did you test it without this bit of code?

File: src/openrct2/actions/StaffSetPatrolAreaAction.cpp [hunk 9]
```diff
@@ -40,11 +48,6 @@ GameActions::Result StaffSetPatrolAreaAction::Query() const
         return GameActions::Result(GameActions::Status::InvalidParameters, STR_NONE, STR_NONE);
     }
 
-    if (!LocationValid(_loc))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/StaffSetPatrolAreaAction.cpp:43]: you still want to run through LocationValid on all locations. Also don't you need to truncate the range if it goes off map?

File: src/openrct2-ui/windows/PatrolArea.cpp (hunk 8 above)
Comment (IntelOrca) [reply src/openrct2-ui/windows/PatrolArea.cpp:282]: It needs to take into account the staff ID is the same. I can change `WindowFocusOrCreate` to just `WindowCreate` if that is less confusing.

File: src/openrct2/actions/StaffSetPatrolAreaAction.cpp (hunk 9 above)
Comment (IntelOrca) [reply src/openrct2/actions/StaffSetPatrolAreaAction.cpp:43]: Are there helper methods for MapRange that I can use?

I don't want to deny the action if just one tile is invalid. Any invalid tiles could be ignored by the Staff / PatrolArea methods.

Comment (duncanspumpkin) [reply src/openrct2/actions/StaffSetPatrolAreaAction.cpp:43]: The location valid is meant to not just check if the location is inside bounds but also pass checking to plugins to allow them to block locations. So you basically want to break it up into an individual tile loop. This is done in the other tools that effect an area as well making it only remove those that are not valid

File: src/openrct2-ui/windows/PatrolArea.cpp (hunk 8 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/PatrolArea.cpp:282]: But don't you force the staff ID in the bit of code after the focus or create

File: src/openrct2-ui/windows/PatrolArea.cpp [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+#include <openrct2/localisation/Formatter.h>
+#include <openrct2/localisation/Localisation.h>
+#include <openrct2/world/Park.h>
+
+static constexpr const rct_string_id WINDOW_TITLE = STR_SET_PATROL_AREA;
+static constexpr const int32_t WH = 54;
+static constexpr const int32_t WW = 104;
+
+// clang-format off
+enum WindowLandRightsWidgetIdx {
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/PatrolArea.cpp:33]: One of the compilers is complaining about reusing names

File: src/openrct2-ui/windows/PatrolArea.cpp [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    WIDX_CLOSE,
+    WIDX_PREVIEW,
+    WIDX_DECREMENT,
+    WIDX_INCREMENT,
+};
+
+static rct_widget PatrolAreaWidgets[] = {
+    WINDOW_SHIM(WINDOW_TITLE, WW, WH),
+    MakeWidget     ({27, 17}, {44, 32}, WindowWidgetType::ImgBtn,  WindowColour::Primary , SPR_LAND_TOOL_SIZE_0                                           ), // preview box
+    MakeRemapWidget({28, 18}, {16, 16}, WindowWidgetType::TrnBtn,  WindowColour::Tertiary, SPR_LAND_TOOL_DECREASE,      STR_ADJUST_SMALLER_LAND_RIGHTS_TIP), // decrement size
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/PatrolArea.cpp:45]: Are these tooltips generic enough?

File: src/openrct2-ui/windows/PatrolArea.cpp (hunk 8 above)
Comment (IntelOrca) [reply src/openrct2-ui/windows/PatrolArea.cpp:282]: I don't know whether it is better to switch the staff ID or to re-create the window which will reset the position of it. What do you think?
